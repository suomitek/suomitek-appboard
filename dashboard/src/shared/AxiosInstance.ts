import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  InternalServerError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response) {
        if (err.response.data.message) {
          message = err.response.data.message;
        }
        if (typeof err.response.data === "string") {
          message = err.response.data;
        }
      }

      const dispatchErrorAndLogout = (m: string) => {
        // Global action dispatch to log the user out
        dispatch(actions.auth.authenticationError(m));
        // Expire the session if we are using OIDC tokens and
        // logout either way.
        dispatch(actions.auth.expireSession());
      };
      const response = err.response as AxiosResponse;
      switch (response && response.status) {
        case 401:
          dispatchErrorAndLogout(message);
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          // A 403 directly from the auth proxy requires reauthentication.
          if (Auth.usingOIDCToken() && Auth.is403FromAuthProxy(response)) {
            dispatchErrorAndLogout(message);
          }
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        case 500:
          return Promise.reject(new InternalServerError(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
export const axiosWithAuth = Axios.create();
