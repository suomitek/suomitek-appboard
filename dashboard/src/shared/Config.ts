import axios from "axios";

interface ICluster {
  name: string;
  apiServiceURL: string;
  certificateAuthorityData?: string;
}

export interface IFeatureFlags {
  operators: boolean;
  ui: string;
}

// IConfig is the configuration for Kubeapps
export interface IConfig {
  namespace: string;
  appVersion: string;
  authProxyEnabled: boolean;
  oauthLoginURI: string;
  oauthLogoutURI: string;
  error?: Error;
  featureFlags: IFeatureFlags;
  clusters: ICluster[];
}

export default class Config {
  public static async getConfig() {
    const url = Config.APIEndpoint;
    const { data } = await axios.get<IConfig>(url);

    // Development environment config overrides
    // TODO(miguel) Rename env variable to KUBEAPPS_NAMESPACE once/if we eject create-react-app
    // Currently we are using REACT_APP_* because it's the only way to inject env variables in a sealed setup.
    // Please note that this env variable gets mapped in the run command in the package.json file
    if (process.env.NODE_ENV !== "production") {
      data.appVersion = "DEVEL";
      if (process.env.REACT_APP_KUBEAPPS_NS) {
        data.namespace = process.env.REACT_APP_KUBEAPPS_NS;
      }
    }

    return data;
  }

  private static APIEndpoint: string = "config.json";
}
