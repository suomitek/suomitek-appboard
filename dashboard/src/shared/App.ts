import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const KUBEOPS_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static async create(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.list(cluster, namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(
    cluster: string,
    namespace: string,
    releaseName: string,
    revision: number,
  ) {
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(
    cluster: string,
    namespace: string,
    releaseName: string,
    purge: boolean,
  ) {
    let endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  public static async listApps(cluster: string, namespace?: string, allStatuses?: boolean) {
    let endpoint = namespace
      ? url.kubeops.releases.list(cluster, namespace)
      : url.kubeops.releases.listAll(cluster);
    if (allStatuses) {
      endpoint += "?statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(endpoint);
    return data.data;
  }

  public static async getRelease(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      url.kubeops.releases.get(cluster, namespace, name),
    );
    return data.data;
  }
}
