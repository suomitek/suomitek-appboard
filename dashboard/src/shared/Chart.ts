import { JSONSchema4 } from "json-schema";
import { axiosWithAuth } from "./AxiosInstance";
import { IChart, IChartVersion } from "./types";
import * as URL from "./url";

export default class Chart {
  public static async fetchCharts(namespace: string, repo: string) {
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(
      URL.api.charts.list(namespace, repo),
    );
    return data.data;
  }

  public static async fetchChartVersions(namespace: string, id: string): Promise<IChartVersion[]> {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion[] }>(
      URL.api.charts.listVersions(namespace, id),
    );
    return data.data;
  }

  public static async getChartVersion(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion }>(
      URL.api.charts.getVersion(namespace, id, version),
    );
    return data.data;
  }

  public static async getReadme(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getReadme(namespace, id, version),
    );
    return data;
  }

  public static async getValues(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getValues(namespace, id, version),
    );
    return data;
  }

  public static async getSchema(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<JSONSchema4>(
      URL.api.charts.getSchema(namespace, id, version),
    );
    return data;
  }

  public static async listWithFilters(
    namespace: string,
    name: string,
    version: string,
    appVersion: string,
  ) {
    const url = `${
      Chart.APIEndpoint
    }/ns/${namespace}/charts?name=${name}&version=${encodeURIComponent(
      version,
    )}&appversion=${appVersion}`;
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/assetsvc/v1";
}
