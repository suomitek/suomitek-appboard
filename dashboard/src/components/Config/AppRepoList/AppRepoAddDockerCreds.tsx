import * as React from "react";

import { ISecret } from "../../../shared/types";

interface IAppRepoFormProps {
  imagePullSecrets: ISecret[];
  togglePullSecret: (imagePullSecret: string) => () => void;
  selectedImagePullSecrets: { [key: string]: boolean };
  namespace: string;
  createDockerRegistrySecret: (
    name: string,
    user: string,
    password: string,
    email: string,
    server: string,
    namespace: string,
  ) => Promise<boolean>;
  fetchImagePullSecrets: (namespace: string) => void;
}

interface IAppRepoFormState {
  user: string;
  password: string;
  email: string;
  server: string;
  secretName: string;
  showSecretSubForm: boolean;
  creating: boolean;
}

export class AppRepoAddDockerCreds extends React.Component<IAppRepoFormProps, IAppRepoFormState> {
  public state: IAppRepoFormState = {
    secretName: "",
    user: "",
    password: "",
    email: "",
    server: "",
    showSecretSubForm: false,
    creating: false,
  };

  public render() {
    const { imagePullSecrets, togglePullSecret, selectedImagePullSecrets } = this.props;
    const { showSecretSubForm } = this.state;
    return (
      <div className="margin-l-big margin-t-normal">
        {imagePullSecrets.length > 0 ? (
          imagePullSecrets.map(secret => {
            return (
              <div key={secret.metadata.name}>
                <label
                  className="checkbox"
                  key={secret.metadata.name}
                  onChange={togglePullSecret(secret.metadata.name)}
                >
                  <input type="checkbox" checked={selectedImagePullSecrets[secret.metadata.name]} />
                  <span>{secret.metadata.name}</span>
                </label>
              </div>
            );
          })
        ) : (
          <div className="margin-b-small">No existing credentials found.</div>
        )}
        {this.state.showSecretSubForm && (
          <div className="secondary-input margin-t-big">
            <div className="row">
              <div className="col-1 margin-t-normal">
                <label htmlFor="kubeapps-docker-cred-secret-name">Secret Name</label>
              </div>
              <div className="col-11">
                <input
                  id="kubeapps-docker-cred-secret-name"
                  value={this.state.secretName}
                  onChange={this.handleSecretNameChange}
                  placeholder="Secret"
                  required={true}
                />
              </div>
            </div>
            <div className="row">
              <div className="col-1 margin-t-normal">
                <label htmlFor="kubeapps-docker-cred-server">Server</label>
              </div>
              <div className="col-11">
                <input
                  id="kubeapps-docker-cred-server"
                  value={this.state.server}
                  onChange={this.handleServerChange}
                  placeholder="https://index.docker.io/v1/"
                  required={true}
                />
              </div>
            </div>
            <div className="row">
              <div className="col-1 margin-t-normal">
                <label htmlFor="kubeapps-docker-cred-username">Username</label>
              </div>
              <div className="col-11">
                <input
                  id="kubeapps-docker-cred-username"
                  value={this.state.user}
                  onChange={this.handleUserChange}
                  placeholder="Username"
                  required={true}
                />
              </div>
            </div>
            <div className="row">
              <div className="col-1 margin-t-normal">
                <label htmlFor="kubeapps-docker-cred-password">Password</label>
              </div>
              <div className="col-11">
                <input
                  type="password"
                  id="kubeapps-docker-cred-password"
                  value={this.state.password}
                  onChange={this.handlePasswordChange}
                  placeholder="Password"
                  required={true}
                />
              </div>
            </div>
            <div className="row">
              <div className="col-1 margin-t-normal">
                <label htmlFor="kubeapps-docker-cred-email">Email</label>
              </div>
              <div className="col-11">
                <input
                  id="kubeapps-docker-cred-email"
                  value={this.state.email}
                  onChange={this.handleEmailChange}
                  placeholder="user@example.com"
                  required={true}
                />
              </div>
            </div>
            <div>
              <button
                className="button button-primary"
                type="button"
                disabled={this.state.creating}
                onClick={this.handleInstallClick}
              >
                {this.state.creating ? "Creating..." : "Submit"}
              </button>
              <button onClick={this.toggleCredSubForm} type="button" className="button">
                Cancel
              </button>
            </div>
          </div>
        )}
        {!showSecretSubForm && (
          <button onClick={this.toggleCredSubForm} className="button margin-t-normal" type="button">
            Add new credentials
          </button>
        )}
      </div>
    );
  }

  private handleUserChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ user: e.target.value });
  };

  private handleSecretNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ secretName: e.target.value });
  };

  private handlePasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ password: e.target.value });
  };

  private handleEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ email: e.target.value });
  };

  private handleServerChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ server: e.target.value });
  };

  private toggleCredSubForm = () => {
    this.setState({ showSecretSubForm: !this.state.showSecretSubForm });
  };

  private handleInstallClick = async () => {
    const { fetchImagePullSecrets, namespace } = this.props;
    const { secretName, user, password, email, server } = this.state;
    const success = await this.props.createDockerRegistrySecret(
      secretName,
      user,
      password,
      email,
      server,
      namespace,
    );
    if (success) {
      // re-fetch secrets
      fetchImagePullSecrets(namespace);
      this.setState({
        secretName: "",
        user: "",
        password: "",
        email: "",
        server: "",
        showSecretSubForm: false,
        creating: false,
      });
    }
  };
}

export default AppRepoAddDockerCreds;
