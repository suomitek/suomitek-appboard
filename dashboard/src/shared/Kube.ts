import * as url from "shared/url";
import { Auth } from "./Auth";
import { axiosWithAuth } from "./AxiosInstance";
import { ResourceKindsWithAPIVersions } from "./ResourceAPIVersion";
import { isNamespaced, ResourceKind, ResourceKindsWithPlurals } from "./ResourceKinds";
import { IK8sList, IResource } from "./types";

export const APIBase = "api/clusters/default";
export let WebSocketAPIBase: string;
if (window.location.protocol === "https:") {
  WebSocketAPIBase = `wss://${window.location.host}${window.location.pathname}`;
} else {
  WebSocketAPIBase = `ws://${window.location.host}${window.location.pathname}`;
}

// Kube is a lower-level class for interacting with the Kubernetes API. Use
// ResourceRef to interact with a single API resource rather than using Kube
// directly.
export class Kube {
  public static getResourceURL(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let u = `${url.api.k8s.base(cluster)}/${
      apiVersion === "v1" || !apiVersion ? "api/v1" : `apis/${apiVersion}`
    }`;
    if (namespace && isNamespaced(resource)) {
      u += `/namespaces/${namespace}`;
    }
    u += `/${resource}`;
    if (name) {
      u += `/${name}`;
    }
    if (query) {
      u += `?${query}`;
    }
    return u;
  }

  public static watchResourceURL(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let u = this.getResourceURL(cluster, apiVersion, resource, namespace);
    u = `${WebSocketAPIBase}${u}?watch=true`;
    if (name) {
      u += `&fieldSelector=metadata.name%3D${name}`;
    }
    if (query) {
      u += `&${query}`;
    }
    return u;
  }

  public static async getResource(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    const { data } = await axiosWithAuth.get<IResource | IK8sList<IResource, {}>>(
      this.getResourceURL(cluster, apiVersion, resource, namespace, name, query),
    );
    return data;
  }

  // Opens and returns a WebSocket for the requested resource. Note: it is
  // important that this socket be properly closed when no longer needed. The
  // returned WebSocket can be attached to an event listener to read data from
  // the socket.
  public static watchResource(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    return new WebSocket(
      this.watchResourceURL(cluster, apiVersion, resource, namespace, name, query),
      Auth.wsProtocols(),
    );
  }

  // Gets the plural form of the resource Kind for use in the resource path
  public static resourcePlural(kind: ResourceKind) {
    return ResourceKindsWithPlurals[kind];
  }

  // Gets the apiVersion of the resource Kind for use in the resource path
  public static resourceAPIVersion(kind: ResourceKind) {
    return ResourceKindsWithAPIVersions[kind];
  }
}
