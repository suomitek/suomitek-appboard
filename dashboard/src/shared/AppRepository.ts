import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace, name));
    return data;
  }

  public static async resync(name: string, namespace: string) {
    const repo = await AppRepository.get(name, namespace);
    repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
    repo.spec.resyncRequests++;
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(namespace, name), repo);
    return data;
  }

  public static async update(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(namespace, name),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(name, namespace),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(namespace),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async validate(repoURL: string, authHeader: string, customCA: string) {
    const { data } = await axiosWithAuth.post<any>(url.backend.apprepositories.validate(), {
      appRepository: { repoURL, authHeader, customCA },
    });
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getSelfLink(namespace: string, name?: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
