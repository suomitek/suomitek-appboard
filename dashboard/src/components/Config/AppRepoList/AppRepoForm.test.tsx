import { shallow } from "enzyme";
import * as React from "react";
import { ISecret } from "shared/types";
import { wait } from "../../../shared/utils";
import UnexpectedErrorPage from "../../ErrorAlert/UnexpectedErrorAlert";
import AppRepoAddDockerCreds from "./AppRepoAddDockerCreds";
import { AppRepoForm } from "./AppRepoForm";

const defaultProps = {
  onSubmit: jest.fn(),
  validate: jest.fn(),
  validating: false,
  imagePullSecrets: [],
  namespace: "default",
  kubeappsNamespace: "kubeapps",
  fetchImagePullSecrets: jest.fn(),
  createDockerRegistrySecret: jest.fn(),
};

it("should render the repo form", () => {
  const wrapper = shallow(<AppRepoForm {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

it("disables the submit button while fetching", () => {
  const wrapper = shallow(<AppRepoForm {...defaultProps} validating={true} />);
  expect(wrapper.find("button").prop("disabled")).toBe(true);
});

it("should show a validation error", () => {
  const wrapper = shallow(<AppRepoForm {...defaultProps} validationError={new Error("Boom!")} />);
  expect(
    wrapper
      .find(UnexpectedErrorPage)
      .dive()
      .text(),
  ).toContain("Boom!");
});

it("should call the install method when the validation success", async () => {
  const validate = jest.fn().mockReturnValue(true);
  const install = jest.fn().mockReturnValue(true);
  const wrapper = shallow(<AppRepoForm {...defaultProps} validate={validate} onSubmit={install} />);
  const button = wrapper.find("form");
  button.simulate("submit", { preventDefault: jest.fn() });
  // wait for async functions
  await new Promise(s => s());
  expect(install).toHaveBeenCalled();
});

it("should not call the install method when the validation fails unless forced", async () => {
  const validate = jest.fn().mockReturnValue(false);
  const install = jest.fn().mockReturnValue(true);
  const wrapper = shallow(<AppRepoForm {...defaultProps} validate={validate} onSubmit={install} />);
  let button = wrapper.find("form");

  button.simulate("submit", { preventDefault: jest.fn() });
  // wait for async functions
  await new Promise(s => s());
  expect(install).not.toHaveBeenCalled();
  wrapper.update();
  button = wrapper.find("button");
  expect(button.text()).toContain("Install Repo (force)");

  wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
  // wait for async functions
  await new Promise(s => s());
  expect(install).toHaveBeenCalled();
});

it("should not show the docker registry credentials section if the namespace is the global one", () => {
  const wrapper = shallow(
    <AppRepoForm {...defaultProps} kubeappsNamespace={defaultProps.namespace} />,
  );
  expect(wrapper.text()).not.toContain("Associate Docker Registry Credentials");
});

it("should render the docker registry credentials section", () => {
  const wrapper = shallow(<AppRepoForm {...defaultProps} />);
  expect(wrapper.find(AppRepoAddDockerCreds)).toExist();
});

it("should call the install method with the selected docker credentials", async () => {
  const validate = jest.fn().mockReturnValue(true);
  const install = jest.fn().mockReturnValue(true);
  const wrapper = shallow(<AppRepoForm {...defaultProps} validate={validate} onSubmit={install} />);
  wrapper.setState({ selectedImagePullSecrets: { "repo-1": true } });

  const button = wrapper.find("form");
  button.simulate("submit", { preventDefault: jest.fn() });
  // wait for async functions
  await wait(1);
  expect(install).toHaveBeenCalledWith("", "", "", "", "", ["repo-1"]);
});

it("should parse and preserve the selection of docker credentials", () => {
  const secret1 = {
    metadata: {
      name: "foo",
    },
  } as ISecret;
  const secret2 = {
    metadata: {
      name: "bar",
    },
  } as ISecret;
  const wrapper = shallow(<AppRepoForm {...defaultProps} imagePullSecrets={[secret1]} />);
  expect(wrapper.state()).toMatchObject({ selectedImagePullSecrets: { foo: false } });

  // Select secret
  wrapper.setState({ selectedImagePullSecrets: { foo: true } });

  // Add new secret, force re-render
  wrapper.setProps({ imagePullSecrets: [secret1, secret2] });

  expect(wrapper.state()).toMatchObject({ selectedImagePullSecrets: { foo: true, bar: false } });
});

describe("when the repository info is already populated", () => {
  it("should parse the existing name", () => {
    const repo = { metadata: { name: "foo" } } as any;
    const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} />);
    expect(wrapper.state()).toMatchObject({ name: "foo" });
    // It should also disable the name input if it's already been set
    expect(
      wrapper
        .find("input")
        .findWhere(i => i.prop("id") === "kubeapps-repo-name")
        .prop("disabled"),
    ).toBe(true);
  });

  it("should parse the existing url", () => {
    const repo = { metadata: { name: "foo" }, spec: { url: "http://repo" } } as any;
    const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} />);
    expect(wrapper.state()).toMatchObject({ url: "http://repo" });
  });

  it("should parse the existing syncJobPodTemplate", () => {
    const repo = { metadata: { name: "foo" }, spec: { syncJobPodTemplate: { foo: "bar" } } } as any;
    const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} />);
    expect(wrapper.state()).toMatchObject({ syncJobPodTemplate: "foo: bar\n" });
  });

  describe("when there is a secret associated to the repo", () => {
    it("should parse the existing CA cert", () => {
      const repo = { metadata: { name: "foo" } } as any;
      const secret = { data: { "ca.crt": "Zm9v" } } as any;
      const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} secret={secret} />);
      expect(wrapper.state()).toMatchObject({ customCA: "foo" });
    });

    it("should parse the existing auth header", () => {
      const repo = { metadata: { name: "foo" } } as any;
      const secret = { data: { authorizationHeader: "Zm9v" } } as any;
      const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} secret={secret} />);
      expect(wrapper.state()).toMatchObject({ authHeader: "foo", authMethod: "custom" });
    });

    it("should parse the existing basic auth", () => {
      const repo = { metadata: { name: "foo" } } as any;
      const secret = { data: { authorizationHeader: "QmFzaWMgWm05dk9tSmhjZz09" } } as any;
      const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} secret={secret} />);
      expect(wrapper.state()).toMatchObject({
        authHeader: "Basic Zm9vOmJhcg==",
        user: "foo",
        password: "bar",
        authMethod: "basic",
      });
    });

    it("should parse a bearer token", () => {
      const repo = { metadata: { name: "foo" } } as any;
      const secret = { data: { authorizationHeader: "QmVhcmVyIGZvbw==" } } as any;
      const wrapper = shallow(<AppRepoForm {...defaultProps} repo={repo} secret={secret} />);
      expect(wrapper.state()).toMatchObject({
        authHeader: "Bearer foo",
        token: "foo",
        authMethod: "bearer",
      });
    });

    it("should pre-select the existing docker registry secret", () => {
      const secret = {
        metadata: {
          name: "foo",
        },
      } as ISecret;
      const repo = { metadata: { name: "foo" }, spec: { dockerRegistrySecrets: ["foo"] } } as any;
      const wrapper = shallow(
        <AppRepoForm {...defaultProps} imagePullSecrets={[secret]} repo={repo} />,
      );
      expect(wrapper.state()).toMatchObject({ selectedImagePullSecrets: { foo: true } });

      // Add new secret, force re-render
      const secret2 = {
        metadata: {
          name: "bar",
        },
      } as ISecret;
      wrapper.setProps({ imagePullSecrets: [secret, secret2] });

      expect(wrapper.state()).toMatchObject({
        selectedImagePullSecrets: { foo: true, bar: false },
      });
    });
  });
});
