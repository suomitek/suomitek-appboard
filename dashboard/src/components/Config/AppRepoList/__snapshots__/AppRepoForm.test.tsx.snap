// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the repo form 1`] = `
<form
  className="container padding-b-bigger"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <div>
        <h2>
          Add an App Repository
        </h2>
      </div>
      <div>
        <label
          htmlFor="kubeapps-repo-name"
        >
          Name:
        </label>
        <input
          disabled={false}
          id="kubeapps-repo-name"
          onChange={[Function]}
          pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
          placeholder="example"
          required={true}
          title="Use lower case alphanumeric characters, '-' or '.'"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          htmlFor="kubeapps-repo-url"
        >
          URL:
        </label>
        <input
          id="kubeapps-repo-url"
          onChange={[Function]}
          placeholder="https://charts.example.com/stable"
          required={true}
          type="url"
          value=""
        />
      </div>
      <div>
        <p
          className="margin-b-small"
        >
          Repository Authorization (optional):
        </p>
        <span
          className="AppRepoInputDescription"
        >
          Introduce the credentials to access the Chart repository if authentication is enabled.
        </span>
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            <label
              className="margin-l-big"
              htmlFor="kubeapps-repo-auth-method-none"
            >
              <input
                checked={true}
                id="kubeapps-repo-auth-method-none"
                name="auth"
                onChange={[Function]}
                type="radio"
                value="none"
              />
              None
              <br />
            </label>
            <label
              htmlFor="kubeapps-repo-auth-method-basic"
            >
              <input
                checked={false}
                id="kubeapps-repo-auth-method-basic"
                name="auth"
                onChange={[Function]}
                type="radio"
                value="basic"
              />
              Basic Auth
              <br />
            </label>
            <label
              htmlFor="kubeapps-repo-auth-method-bearer"
            >
              <input
                checked={false}
                id="kubeapps-repo-auth-method-bearer"
                name="auth"
                onChange={[Function]}
                type="radio"
                value="bearer"
              />
              Bearer Token
              <br />
            </label>
            <label
              htmlFor="kubeapps-repo-auth-method-custom"
            >
              <input
                checked={false}
                id="kubeapps-repo-auth-method-custom"
                name="auth"
                onChange={[Function]}
                type="radio"
                value="custom"
              />
              Custom
              <br />
            </label>
          </div>
          <div
            aria-live="polite"
            className="col-10"
          >
            <div
              className="secondary-input"
              hidden={true}
            >
              <label
                htmlFor="kubeapps-repo-username"
              >
                Username
              </label>
              <input
                id="kubeapps-repo-username"
                onChange={[Function]}
                placeholder="Username"
                type="text"
                value=""
              />
              <label
                htmlFor="kubeapps-repo-password"
              >
                Password
              </label>
              <input
                id="kubeapps-repo-password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              className="secondary-input"
              hidden={true}
            >
              <label
                htmlFor="kubeapps-repo-token"
              >
                Token
              </label>
              <input
                id="kubeapps-repo-token"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="secondary-input"
              hidden={true}
            >
              <label
                htmlFor="kubeapps-repo-custom-header"
              >
                Complete Authorization Header
              </label>
              <input
                id="kubeapps-repo-custom-header"
                onChange={[Function]}
                placeholder="Bearer xrxNcWghpRLdcPHFgVRM73rr4N7qjvjm"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div>
        <p
          className="margin-b-small"
        >
          Associate Docker Registry Credentials (optional):
        </p>
        <span
          className="AppRepoInputDescription"
        >
          Select existing secret(s) to access a private Docker registry and pull images from it. Note that this functionality is supported for Kubeapps with Helm3 only, more info
           
          <a
            href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/private-app-repository.md"
            rel="noopener noreferrer"
            target="_blank"
          >
            here
          </a>
          .
          <AppRepoAddDockerCreds
            createDockerRegistrySecret={[MockFunction]}
            fetchImagePullSecrets={[MockFunction]}
            imagePullSecrets={Array []}
            namespace="default"
            selectedImagePullSecrets={Object {}}
            togglePullSecret={[Function]}
          />
        </span>
      </div>
      <div
        className="margin-t-big"
      >
        <label>
          <span>
            Custom CA Certificate (optional):
          </span>
          <pre
            className="CodeContainer"
          >
            <textarea
              className="Code"
              onChange={[Function]}
              placeholder="-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----"
              rows={4}
              value=""
            />
          </pre>
        </label>
      </div>
      <div>
        <label
          htmlFor="syncJobPodTemplate"
        >
          Custom Sync Job Template (optional)
        </label>
        <Hint
          id="syncJobHelp"
          reactTooltipOpts={
            Object {
              "delayHide": 1000,
            }
          }
        >
          <span>
            It's possible to modify the default sync job. More info
             
            <a
              href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/private-app-repository.md#modifying-the-synchronization-job"
              rel="noopener noreferrer"
              target="_blank"
            >
              here
            </a>
            <br />
            When modifying the default sync job, the pre-validation is not supported.
          </span>
        </Hint>
        <pre
          className="CodeContainer"
        >
          <textarea
            className="Code"
            id="syncJobPodTemplate"
            onChange={[Function]}
            placeholder="spec:
  containers:
  - env:
    - name: FOO
      value: BAR
"
            rows={4}
            value=""
          />
        </pre>
      </div>
      <div>
        <button
          className="button button-primary"
          disabled={false}
          type="submit"
        >
          Install Repo 
        </button>
      </div>
    </div>
  </div>
</form>
`;
